api/metadata/serializers.py

```diff
@@ -0,0 +1,133 @@
+                f"Value string is too long. Must be less than {FIELD_VALUE_MAX_LENGTH} character"
+            )
+        return data
+
+    def to_representation(self, value):
+        return value
+
+
+class MetadataListSerializer(serializers.ListSerializer):
+    def save(self, **kwargs):
+        validated_data = [{**attrs, **kwargs} for attrs in self.validated_data]
+        content_object = kwargs["content_object"]
+        content_type = ContentType.objects.get_for_model(content_object)
+        instances = []
+
+        # Create or update the metadata
+        for metadata in validated_data:
+            instance, _ = Metadata.objects.update_or_create(
+                content_type=content_type,
+                object_id=content_object.id,
+                model_field=metadata["model_field"],
+                defaults={"field_value": metadata["field_value"]},
+            )
+            instances.append(instance)
+
+        # Delete the metadata not in the list
+        Metadata.objects.filter(
+            ~Q(id__in=[instance.id for instance in instances]),
+            content_type=content_type,
+            object_id=content_object.id,
+        ).delete()
+
+        return instances
+
+
+class MetadataSerializer(serializers.ModelSerializer):
+    field_value = FieldDataField()
+
+    class Meta:
+        model = Metadata
+        fields = ("model_field", "field_value")
+        list_serializer_class = MetadataListSerializer
+
+    def validate(self, data):
+        if type(data["field_value"]).__name__ != data["model_field"].field.type:
+            raise serializers.ValidationError("Invalid type")
+        return data
+
+    def to_representation(self, value):
+        # Convert field_value to its appropriate type(from string)
+        field_type = value.model_field.field.type
+        value = super().to_representation(value)
+
+        value["field_value"] = {
+            "str": str,
+            "int": int,
+            "float": float,
+            "bool": lambda v: v not in ("False", "false"),
+        }.get(field_type)(value["field_value"])
+
+        return value
+
+
+class MetadataSerializerMixin:
+    def validate_required_metadata(self):
+        metadata = self.validated_data.pop("metadata", [])
+
+        content_type = ContentType.objects.get_for_model(self.Meta.model)
+
+        required_metadata_fields = MetadataModelField.objects.filter(
+            content_type=content_type, is_required=True
+        )
+
+        for metadata_field in required_metadata_fields:
+            if not any([field["model_field"] == metadata_field for field in metadata]):
+                raise serializers.ValidationError(
+                    f"Missing required metadata field {metadata_field.field.name}"
+                )
+
+    def save(self, **kwargs):
+        self.validate_required_metadata()
+
+        instance = super().save(**kwargs)
+
+        metadata = self.get_initial().get("metadata", None)
```

Comment:
Can we use a WritableNestedSerializer for this instead? or perhaps not because of all the GenericRelation stuff?

Reply:
Hmm, I think there was some reason (Ah no it was related to the old design) good shout